### PR TITLE
add documentation for distributed query and using predicates

### LINKED ISSUE #261: Under documented predicates
There is no documentation around using Predicates on Node.js side. Particularly, it is not possible for users to know how to use custom comparators by looking at the documentation.
See https://github.com/hazelcast/hazelcast-nodejs-client/issues/255

### PR DESCRIPTION
**Pull request by furkansenharputlu**
This PR adds documentation for distributed query and using predicates. 
It also explains how to add user library to server `CLASSPATH`.
fixes #261 
#340

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1458,6 +1479,233 @@ map.addEntryListener(mapEventListener).then(function () {
...
+
+The `Predicates` object offered by the client includes many operators for your query requirements. Some of them are explained below.
+
+- `equal`: Checks if the result of an expression is equal to a given value.
+
+- `notEqual`: Checks if the result of an expression is not equal to a given value.
+
+- `instanceOf`: Checks if the result of an expression has a certain type.
+
+- `like`: Checks if the result of an expression matches some string pattern. % (percentage sign) is the placeholder for many characters, (underscore) is placeholder for only one character.
```
Comment (mustafaiman) [inline README.md]:
```suggestion
- `like`: Checks if the result of an expression matches some string pattern. `%` (percentage sign) is the placeholder for many characters, `_` (underscore) is placeholder for only one character.
```

Comment (mustafaiman) [inline README.md]:
```suggestion
- `like`: Checks if the result of an expression matches some string pattern. `%` (percentage sign) is the placeholder for many characters, `_` (underscore) is placeholder for only one character.
```

File: README.md [hunk 2]
```diff
@@ -1458,6 +1479,233 @@ map.addEntryListener(mapEventListener).then(function () {
...
+- `not`: Checks if the result of an expression is false.
+
+- `regex`: Checks if the result of an expression matches some regular expression.
+
+#### 1.3. Querying with SQL
+
+`SqlPredicate` takes the regular SQL `where` clause. Here is an example:
+
+```javascript
+var map = client.getMap('employee');
```
Comment (mustafaiman) [inline README.md]: old API

File: README.md [hunk 3]
```diff
@@ -1458,6 +1479,233 @@ map.addEntryListener(mapEventListener).then(function () {
...
+Similar to LIKE predicate but in a case-insensitive manner.
+
+- `name ILIKE 'Jo%'` (true for 'Joe', 'joe', 'jOe','Josh','joSH', etc.)
+- `name ILIKE 'Jo_'` (true for 'Joe' or 'jOE'; false for 'Josh')
+
+**REGEX:** `<attribute> [NOT] REGEX 'expression'`
+
+- `name REGEX 'abc-.*'` (true for 'abc-123'; false for 'abx-123')
+
+##### Querying Entry Keys with Predicates
```
Comment (mustafaiman) [inline README.md]: There is also `_this` keyword for values. Why don't we talk about it as well?

File: README.md [hunk 4]
```diff
@@ -1458,6 +1479,233 @@ map.addEntryListener(mapEventListener).then(function () {
...
+    pagingPredicate.nextPage();
+    
+    // Retrieve next page
+    return map.valuesWithPredicate(pagingPredicate);
+}).then(function (values) {
+    // some operations
+});
+```
+
+If you want to sort the result before paging, you need to specify a comparator object that implements `Comparator` interface. Also, this comparator object should be one of `IdentifiedDataSerializable` or `Portable`. After implementing the Node.js version, you need to implement the Java equivalent of the comparator and its factory. The Java equivalent of the comparator should implement `java.util.Comparator` interface's `compare` method which will specify how to sort the result. Note that `comprare` function of `Comparator` on the Java side is the equivalent of the `sort` function of `Comparator` on the Node.js side. When you implement `Comparator` and its factory, you can add them to the `CLASSPATH` of the server side.  See [Adding User Library to CLASSPATH section](#adding-user-library-to-classpath).
```
Comment (mustafaiman) [inline README.md]: "..should implement `java.util.Comparator` interface's `compare` method which will specify how to sort the result." -> ".. should implement `java.util.Comparator`"

File: README.md [hunk 5]
```diff
@@ -1458,6 +1479,233 @@ map.addEntryListener(mapEventListener).then(function () {
...
+
+Also, You can access a specific page more easily with the help of the `setPage` function. This way, if you make a query for the hundredth page, for example, it will get all 100 pages at once instead of reaching the hundredth page one by one using the `nextPage` function.
+
```
Comment (mustafaiman) [inline README.md]: AFAIK, paging is beneficial when you have huge number of items. The servers do not have to bring all the items in memory at once. I think we should mention that too

File: README.md [hunk 6]
```diff
@@ -1458,6 +1479,233 @@ map.addEntryListener(mapEventListener).then(function () {
...
+3. The predicate requester merges all the results coming from each member into a single set.
+
+Distributed query is highly scalable. If you add new members to the cluster, the partition count for each member is reduced and thus the time spent by each member on iterating its entries is reduced. In addition, the pool of partition threads evaluates the entries concurrently in each member, and the network traffic is also reduced since only filtered data is sent to the requester.
+
+Hazelcast offers the following ways for distributed query purposes:
+
+- Combining Predicates with AND, OR, NOT
+
+- Distributed SQL Query
+
```
[the remark this answers is not in the document]
Comment (furkansenharputlu) [reply README.md]: They are in the first way. It means `Combining Predicates(Between, GreaterThan, LessThan) with AND, OR, NOT`.

File: README.md [hunk 7]
```diff
@@ -1635,6 +1659,240 @@ map.addEntryListener(mapEventListener).then(function () {
...
+}).then(function () {
+    var predicate = new Predicates.sql('__key like A%');
+    return personMap.valuesWithPredicate(predicate);
+}).then(function (startingWithA) {
+    // some operations
+});
+```
+
+In this example, the code creates a list with the values whose keys start with the letter "A”.
+
```
Comment (mustafaiman) [inline README.md]:
```suggestion

You can also use `this` keyword to use the value of the entry in your predicates.

```javascript
var personMap = client.getMap('persons');
personMap.put('Ahmet', new Person('Ahmet', 27)).then(function () {
    return personMap.put('Ali', new Person('Ali', 30));
}).then(function () {
    return personMap.put('Furkan', new Person('Furkan', 23));
}).then(function () {
    var predicate = new Predicates.greaterThan('this', 27));
    return personMap.valuesWithPredicate(predicate);
}).then(function (olderThan27) {
    // some operations
});
```
```

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mustafaiman) [reply README.md]: I think we need to introduce the predicates before telling that the predicates are combinable. Right now, I read `Combining Predicates with AND, OR, NOT` and say "What am I going to combine with these?"

File: README.md [hunk 8]
```diff
@@ -1635,6 +1659,267 @@ map.addEntryListener(mapEventListener).then(function () {
...
+            </data-serializable-factory>
+        </data-serializable-factories>
+    </serialization>
+    ...
+</hazelcast>
+```
+
+Note that before starting the server, you need to compile `Employee` and `MyIdentifiedFactory` classes with server's `CLASSPATH` and add them to `user-lib` folder in the extracted `hazelcast-<version>.zip`. See [Adding User Library to CLASSPATH section](#adding-user-library-to-classpath).
+
+> **NOTE: You can also make this object `Portable` and implement its Java equivalent and its portable factory on server side. Note that querying with `Portable` object is faster than `IdentifiedDataSerializable`.**
```
Comment (Serdaro) [inline README.md]: ... on the server side. Note that querying with `Portable` object is faster as compared to `IdentifiedDataSerializable`.

File: README.md [hunk 9]
```diff
@@ -1635,6 +1659,267 @@ map.addEntryListener(mapEventListener).then(function () {
...
+
+#### 7.7.1.2. Querying by Combining Predicates with AND, OR, NOT
+
+You can combine predicates by using the and, or, and not operators, as shown in the below example.
```
Comment (Serdaro) [inline README.md]: ... by using `and`, `or` and `not` operators ...

File: README.md [hunk 10]
```diff
@@ -1635,6 +1659,267 @@ map.addEntryListener(mapEventListener).then(function () {
...
+    pagingPredicate.nextPage();
+
+    // Retrieve next page
+    return map.valuesWithPredicate(pagingPredicate);
+}).then(function (values) {
+    // some operations
+});
+```
+
+If you want to sort the result before paging, you need to specify a comparator object that implements `Comparator` interface. Also, this comparator object should be one of `IdentifiedDataSerializable` or `Portable`. After implementing the Node.js version, you need to implement the Java equivalent of the comparator and its factory. The Java equivalent of the comparator should implement `java.util.Comparator`. Note that `comprare` function of `Comparator` on the Java side is the equivalent of the `sort` function of `Comparator` on the Node.js side. When you implement `Comparator` and its factory, you can add them to the `CLASSPATH` of the server side.  See [Adding User Library to CLASSPATH section](#adding-user-library-to-classpath).
```
Comment (Serdaro) [inline README.md]: ... implements the `Comparator` interface...

... Note that the `compare` function of `Comparator` on ...
(note the typo "comprare" is fixed here too)

File: README.md [hunk 11]
```diff
@@ -1635,6 +1659,267 @@ map.addEntryListener(mapEventListener).then(function () {
...
+            <data-serializable-factory factory-id="1">
+                mypackage.MyIdentifiedFactory
+            </data-serializable-factory>
+        </data-serializable-factories>
+    </serialization>
+    ...
+</hazelcast>
+```
+
+Note that before starting the server, you need to compile `Employee` and `MyIdentifiedFactory` classes with server's `CLASSPATH` and add them to `user-lib` folder in the extracted `hazelcast-<version>.zip`. See [Adding User Library to CLASSPATH section](#adding-user-library-to-classpath).
```
Comment (Serdaro) [inline README.md]: ... you need to compile the `Employee` and ...

... and add them to the `user-lib` folder...

**Pull request merged**